Bonsai.Design/ObjectTextVisualizer.cs

```diff
@@ -30,25 +31,99 @@ public class ObjectTextVisualizer : BufferedVisualizer
-            if (values.Count > 0)
+            if (values.Count == 0)
+                return;
+
+            var sb = new StringBuilder();
+
+            // Trim old values if bufferSize was reduced
+            while (buffer.Count >= bufferSize)
+                buffer.Dequeue();
+
+            // Add new values to the buffer (and only the ones which might appear)
+            for (int i = Math.Max(0, values.Count - bufferSize); i < values.Count; i++)
             {
-                base.ShowBuffer(values);
-                textBox.Text = string.Join(Environment.NewLine, buffer);
-                textPanel.Invalidate();
+                sb.Clear();
+                AppendDisplayText(sb, values[i]);
+                AppendString(sb.ToString());
             }
+
+            // Update the visual representation of the buffer
+            RefreshVisualization(sb);
         }
 
         /// <inheritdoc/>
         public override void Show(object value)
         {
-            value ??= string.Empty;
-            var text = value.ToString();
-            text = Regex.Replace(text, @"\r|\n", string.Empty);
-            buffer.Enqueue(text);
-            while (buffer.Count > bufferSize)
+            // Updates to this visualizer are expected to go through ShowBuffer
+            Debug.Fail($"Likely unintentional call to {nameof(ObjectTextVisualizer)}.{nameof(Show)}");
+
+            var stringBuilder = new StringBuilder();
+            AppendDisplayText(stringBuilder, value);
+            AppendString(stringBuilder.ToString());
+            RefreshVisualization(stringBuilder);
+        }
+
+        /// <summary>
+        /// Appends the display text for the specified object to the text buffer.
+        /// </summary>
+        /// <param name="stringBuilder">The string builder which receives the display text.</param>
+        /// <param name="value">The object for which to retrieve the display text.</param>
+        protected virtual void AppendDisplayText(StringBuilder stringBuilder, object value)
+        {
+            string rawText = value?.ToString() ?? string.Empty;
+            stringBuilder.EnsureCapacity(stringBuilder.Length + rawText.Length);
+
+            foreach (var c in rawText)
             {
+                switch (c)
+                {
+                    // Carriage returns are presumed to be followed by line feeds, skip them entirely
+                    case '\r':
+                        continue;
+                    // Newlines become space so that things like multi-line JSON or matricies are still visible
+                    case '\n':
+                    case '\x0085': // Next line character (NEL)
+                    case '\x2028': // Unicode line separator
+                    case '\x2029': // Unicode paragraph separator
+                        stringBuilder.Append(' ');
+                        break;
+                    case '\t':
+                        stringBuilder.Append(c);
+                        break;
```

Comment:
Can we move this case next to `default`, or will it break the switch logic? Although I guess we could probably also leave it as it is for readability.

Reply:
It needs to come before the control characters block since tabs are control characters. I think it's better to show the edge case is explicitly handled.